scripts/storage-emulator-integration/conformance/gcs.endpoints.test.ts

```diff
@@ -0,0 +1,261 @@
+        const fileName = "test_upload.jpg";
+        const uploadUrl = await supertest(storageHost)
+          .post(`/v0/b/${storageBucket}/o?name=${fileName}&uploadType=resumable`)
+          .send({})
+          .set({
+            Authorization: "Bearer owner",
+            "X-Goog-Upload-Protocol": "resumable",
+            "X-Goog-Upload-Command": "start",
+          })
+          .expect(200)
+          .then((res) => {
+            return new URL(res.header["x-goog-upload-url"]);
+          });
+
+        const finalizeStatus = await supertest(storageHost)
+          .post(uploadUrl.pathname + uploadUrl.search)
+          .send({})
+          .set({
+            Authorization: "Bearer owner",
+            "X-Goog-Upload-Protocol": "resumable",
+            "X-Goog-Upload-Command": "finalize",
+          })
+          .expect(200)
+          .then((res) => res.header["x-goog-upload-status"]);
+        expect(finalizeStatus).to.equal("final");
+      });
+
+      it("should handle resumable upload with name only in metadata", async () => {
+        const fileName = "test_upload.jpg";
+        const uploadURL = await supertest(storageHost)
+          .post(`/upload/storage/v1/b/${storageBucket}/o?uploadType=resumable`)
+          .send({ name: fileName })
+          .set({
+            Authorization: "Bearer owner",
+          })
+          .expect(200)
+          .then((res) => new URL(res.header["location"]));
+        expect(uploadURL.searchParams?.get("name")).to.equal(fileName);
+      });
+
+      it("should handle multipart upload with name only in metadata", async () => {
+        const body = Buffer.from(`--b1d5b2e3-1845-4338-9400-6ac07ce53c1e\r
+content-type: application/json\r
+\r
+{"name":"test_upload.jpg"}\r
+--b1d5b2e3-1845-4338-9400-6ac07ce53c1e\r
+content-type: text/plain\r
+\r
+hello there!
+\r
+--b1d5b2e3-1845-4338-9400-6ac07ce53c1e--\r
+`);
+        const fileName = "test_upload.jpg";
+        const responseName = await supertest(storageHost)
+          .post(`/upload/storage/v1/b/${storageBucket}/o?uploadType=multipart`)
+          .send(body)
+          .set({
+            Authorization: "Bearer owner",
+            "content-type": "multipart/related; boundary=b1d5b2e3-1845-4338-9400-6ac07ce53c1e",
+          })
+          .expect(200)
+          .then((res) => res.body.name);
+        expect(responseName).to.equal(fileName);
+      });
+
+      it("should return an error message when uploading a file with invalid metadata", async () => {
+        const fileName = "test_upload.jpg";
+        const errorMessage = await supertest(storageHost)
+          .post(`/upload/storage/v1/b/${storageBucket}/o?name=${fileName}`)
+          .set({ Authorization: "Bearer owner", "X-Upload-Content-Type": "foo" })
+          .expect(400)
+          .then((res) => res.body.error.message);
+
+        expect(errorMessage).to.equal("Invalid Content-Type: foo");
+      });
+    });
+  });
+
+  describe(".file()", () => {
```

Comment:
Would be good to flesh out this file as a whole. I think previously we were just using the js sdk/endpoints whenever one was more convenient so now there are lots of holes in the coverage. Could you add a TODO to flesh out the endpoints testing?

Reply:
Done